File: src/types/transfer.ts
import type { TransferDirection, TransferProtocol, TransferState } from './connection';

export type { TransferDirection, TransferProtocol, TransferState };

export interface TransferTask {
  id: string;
  connection_id: string;
  protocol: TransferProtocol;
  direction: TransferDirection;
  local_path: string;
  remote_path: string;
  file_name: string;
  file_size: number;
  transferred_bytes: number;
  state: TransferState;
  speed_bytes_per_sec: number;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface TransferProgressPayload {
  task_id: string;
  file_name: string;
  total_bytes: number;
  transferred_bytes: number;
  speed_bytes_per_sec: number;
  percent: number;
  direction: TransferDirection;
}

export interface SftpFileEntry {
  name: string;
  path: string;
  is_directory: boolean;
  is_symlink: boolean;
  size: number;
  modified: string;
  permissions: string;
  owner: string | null;
  group: string | null;
}

export interface FtpFileEntry {
  name: string;
  path: string;
  is_directory: boolean;
  size: number;
  modified: string;
  permissions: string;
}

File: src-tauri/src/models/transfer.rs
use serde::{Deserialize, Serialize};

/// Transfer direction
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum TransferDirection {
    Upload,
    Download,
}

/// Transfer type (SFTP or FTP)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum TransferProtocol {
    Sftp,
    Ftp,
}

/// Transfer task state
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum TransferState {
    Queued,
    Transferring,
    Paused,
    Completed,
    Failed(String),
    Cancelled,
}

/// A single transfer task
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransferTask {
    pub id: String,
    pub connection_id: String,
    pub protocol: TransferProtocol,
    pub direction: TransferDirection,
    pub local_path: String,
    pub remote_path: String,
    pub file_name: String,
    pub file_size: u64,
    pub transferred_bytes: u64,
    pub state: TransferState,
    pub speed_bytes_per_sec: f64,
    pub created_at: String,
    pub started_at: Option<String>,
    pub completed_at: Option<String>,
}

/// Transfer progress event payload
#[derive(Debug, Clone, Serialize)]
pub struct TransferProgressPayload {
    pub task_id: String,
    pub file_name: String,
    pub total_bytes: u64,
    pub transferred_bytes: u64,
    pub speed_bytes_per_sec: f64,
    pub percent: f32,
    pub direction: TransferDirection,
}
